proxy/src/transparency/client.rs

```diff
@@ -174,7 +174,7 @@ where
                 let is_body_empty = req.body().is_end_stream();
                 let mut req = hyper::Request::from(req.map(BodyStream::new));
                 if is_body_empty {
-                    req.headers_mut().set(hyper::header::ContentLength(0));
+                    req.body_mut().take();
```

Comment:
Let's add a test for a request like this:

```
POST /foo HTTP/1.1
Content-Length: 0
```

In particular, let's verify that the `Content-Length: 0` from the input request is preserved in the output request.

Also, please document why we're doing this (the body is a fake placeholder that is used to workaround the fact that library X doesn't directly support messages without bodies), similar to what you wrote in your recent GitHub comment.

Reply:
BTW, it is confusing why we use req.body().is_end_stream()` here but then below we use `is_empty()`.

Reply:
The two different methods are required because the types are different. At this function, we've received an `http::Request<transparency::HttpBody>`.

Down below, `HttpBody::is_end_stream` is being implemented in terms of `Either<hyper::Body, tower_h2::RecvBody>`.

Reply:
> In particular, let's verify that the Content-Length: 0 from the input request is preserved in the output request.

What exactly are we trying to test here? Something related to transfer-encoding?

FWIW, I don't believe the `Content-Length: 0` header will survive all the way to the end server. hyper also knows that if the body is `None` internally, and thus no body, that it can remove any set `Content-Length` and `Transfer-Encoding` headers, since a client request with neither has the same semantics.

Reply:
There is a difference between an empty body (Content-Length: 0, chunked encoding of a zero-length body) and no body, and we should preserve that distinction. If we already weren't, then we can just file an issue to track fixing that. If we were already maintaining that distinction but wouldn't with this PR, then we should fix this PR to keep maintaining the distinction.

> What exactly are we trying to test here? Something related to transfer-encoding?

The test is to ensure that we don't strip the Conrtent-Length header from a request with an empty body, turning it into a request with no body.

(Rightly or wrongly, some servers treat a empty body differently than a zero-length body. For example, some versions of the HTTP spec say that a POST request must have a body, and some implementations interpret that to mean that there must always be a Content-Length or Transfer-Encoding indicating a body, even if zero-length.)

Reply:
I see what you're saying. I haven't received a single issue filed on hyper about this case, so I do wonder how much it might happen, but I settled on the reasoning that we are a "transparent" proxy, and so not to repair or alter requests other than we absolutely have to.

As to the test, indeed, before, a `Content-Length: 0` was forwarded on, since hyper saw the set body and therefore didn't touch the headers, and after with this patch, hyper cleaned up. So, I've added a check for the existence of a `Content-Length` header, and if it exists, the body is left alone.

Also added comments explaining why this is all being done. For Future Us.